database/src/main/java/com/firebase/ui/database/paging/DatabasePagingSource.java

```diff
@@ -0,0 +1,120 @@
+package com.firebase.ui.database.paging;
+
+import android.annotation.SuppressLint;
+
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.Query;
+
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import androidx.annotation.NonNull;
+import androidx.paging.PagingState;
+import androidx.paging.rxjava3.RxPagingSource;
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+
+public class DatabasePagingSource extends RxPagingSource<String, DataSnapshot> {
+    private final Query mQuery;
+
+    private static final String STATUS_DATABASE_NOT_FOUND = "DATA_NOT_FOUND";
+    private static final String MESSAGE_DATABASE_NOT_FOUND = "Data not found at given child path!";
+    private static final String DETAILS_DATABASE_NOT_FOUND = "No data was returned for the given query: ";
+
+    public DatabasePagingSource(Query query) {
+        this.mQuery = query;
+    }
+
+    /**
+     * DatabaseError.fromStatus() is not meant to be public.
+     */
+    @SuppressLint("RestrictedApi")
+    @NonNull
+    @Override
+    public Single<LoadResult<String, DataSnapshot>> loadSingle(@NonNull LoadParams<String> params) {
+        Task<DataSnapshot> task;
+        if (params.getKey() == null) {
+            task = mQuery.limitToFirst(params.getLoadSize()).get();
+        } else {
+            task = mQuery.startAt(null, params.getKey()).limitToFirst(params.getLoadSize() + 1).get();
+        }
+
+        return Single.fromCallable(() -> {
+            Tasks.await(task);
```

Comment:
I think this will throw if the Tasks fails:
https://developers.google.com/android/reference/com/google/android/gms/tasks/Tasks#await(com.google.android.gms.tasks.Task%3CTResult%3E)

So we can't rely on getting to the next line. Although I think it's fine since you re-throw at the bottom, but I just don't think you'll ever get that far.

Also we can probably remove the `task.isSuccessful` nesting since we know if we even get to the next line, the task is successful.

Reply:
@samtstern Interesting, I didn't know that.

I have wrapped it around a try/catch so that developers using FirebaseUI can get the original cause  instead of `await`'s `ExecutionException`.